# creating first time stamp 10, nov to 14 nov

In [56]:
import pandas as pd
  
per1 = pd.date_range(start ='11-10-2022', 
         end ='11-15-2022', freq ='215S')
  
for val in per1:
    print(val)

2022-11-10 00:00:00
2022-11-10 00:03:35
2022-11-10 00:07:10
2022-11-10 00:10:45
2022-11-10 00:14:20
2022-11-10 00:17:55
2022-11-10 00:21:30
2022-11-10 00:25:05
2022-11-10 00:28:40
2022-11-10 00:32:15
2022-11-10 00:35:50
2022-11-10 00:39:25
2022-11-10 00:43:00
2022-11-10 00:46:35
2022-11-10 00:50:10
2022-11-10 00:53:45
2022-11-10 00:57:20
2022-11-10 01:00:55
2022-11-10 01:04:30
2022-11-10 01:08:05
2022-11-10 01:11:40
2022-11-10 01:15:15
2022-11-10 01:18:50
2022-11-10 01:22:25
2022-11-10 01:26:00
2022-11-10 01:29:35
2022-11-10 01:33:10
2022-11-10 01:36:45
2022-11-10 01:40:20
2022-11-10 01:43:55
2022-11-10 01:47:30
2022-11-10 01:51:05
2022-11-10 01:54:40
2022-11-10 01:58:15
2022-11-10 02:01:50
2022-11-10 02:05:25
2022-11-10 02:09:00
2022-11-10 02:12:35
2022-11-10 02:16:10
2022-11-10 02:19:45
2022-11-10 02:23:20
2022-11-10 02:26:55
2022-11-10 02:30:30
2022-11-10 02:34:05
2022-11-10 02:37:40
2022-11-10 02:41:15
2022-11-10 02:44:50
2022-11-10 02:48:25
2022-11-10 02:52:00
2022-11-10 02:55:35


In [57]:
len(per1)

2010

In [58]:
df=pd.DataFrame()
print(len(df))
import pandas as pd
# create data frame
df = pd.DataFrame({'Time_Stamp': per1 })
  


0


In [59]:
df.head

<bound method NDFrame.head of               Time_Stamp
0    2022-11-10 00:00:00
1    2022-11-10 00:03:35
2    2022-11-10 00:07:10
3    2022-11-10 00:10:45
4    2022-11-10 00:14:20
...                  ...
2005 2022-11-14 23:44:35
2006 2022-11-14 23:48:10
2007 2022-11-14 23:51:45
2008 2022-11-14 23:55:20
2009 2022-11-14 23:58:55

[2010 rows x 1 columns]>

# Creating second dataframe 15,nov to 18 nov

In [28]:
import pandas as pd

per2 = pd.date_range(start ='11-15-2022', 
         end ='11-19-2022', freq ='97S')
  
for val in per2:
    print(val)
print(len(per2))

2022-11-15 00:00:00
2022-11-15 00:01:37
2022-11-15 00:03:14
2022-11-15 00:04:51
2022-11-15 00:06:28
2022-11-15 00:08:05
2022-11-15 00:09:42
2022-11-15 00:11:19
2022-11-15 00:12:56
2022-11-15 00:14:33
2022-11-15 00:16:10
2022-11-15 00:17:47
2022-11-15 00:19:24
2022-11-15 00:21:01
2022-11-15 00:22:38
2022-11-15 00:24:15
2022-11-15 00:25:52
2022-11-15 00:27:29
2022-11-15 00:29:06
2022-11-15 00:30:43
2022-11-15 00:32:20
2022-11-15 00:33:57
2022-11-15 00:35:34
2022-11-15 00:37:11
2022-11-15 00:38:48
2022-11-15 00:40:25
2022-11-15 00:42:02
2022-11-15 00:43:39
2022-11-15 00:45:16
2022-11-15 00:46:53
2022-11-15 00:48:30
2022-11-15 00:50:07
2022-11-15 00:51:44
2022-11-15 00:53:21
2022-11-15 00:54:58
2022-11-15 00:56:35
2022-11-15 00:58:12
2022-11-15 00:59:49
2022-11-15 01:01:26
2022-11-15 01:03:03
2022-11-15 01:04:40
2022-11-15 01:06:17
2022-11-15 01:07:54
2022-11-15 01:09:31
2022-11-15 01:11:08
2022-11-15 01:12:45
2022-11-15 01:14:22
2022-11-15 01:15:59
2022-11-15 01:17:36
2022-11-15 01:19:13


In [29]:
len(per2)

3563

In [30]:
# create data frame df2
df2 = pd.DataFrame({'Time_Stamp': per2 })

# create data frame df3 from 19-11 to 22-11

In [52]:
import pandas as pd

per3 = pd.date_range(start ='11-19-2022', 
         end ='11-23-2022', freq ='74S')
  
for val in per3:
    print(val)

2022-11-19 00:00:00
2022-11-19 00:01:14
2022-11-19 00:02:28
2022-11-19 00:03:42
2022-11-19 00:04:56
2022-11-19 00:06:10
2022-11-19 00:07:24
2022-11-19 00:08:38
2022-11-19 00:09:52
2022-11-19 00:11:06
2022-11-19 00:12:20
2022-11-19 00:13:34
2022-11-19 00:14:48
2022-11-19 00:16:02
2022-11-19 00:17:16
2022-11-19 00:18:30
2022-11-19 00:19:44
2022-11-19 00:20:58
2022-11-19 00:22:12
2022-11-19 00:23:26
2022-11-19 00:24:40
2022-11-19 00:25:54
2022-11-19 00:27:08
2022-11-19 00:28:22
2022-11-19 00:29:36
2022-11-19 00:30:50
2022-11-19 00:32:04
2022-11-19 00:33:18
2022-11-19 00:34:32
2022-11-19 00:35:46
2022-11-19 00:37:00
2022-11-19 00:38:14
2022-11-19 00:39:28
2022-11-19 00:40:42
2022-11-19 00:41:56
2022-11-19 00:43:10
2022-11-19 00:44:24
2022-11-19 00:45:38
2022-11-19 00:46:52
2022-11-19 00:48:06
2022-11-19 00:49:20
2022-11-19 00:50:34
2022-11-19 00:51:48
2022-11-19 00:53:02
2022-11-19 00:54:16
2022-11-19 00:55:30
2022-11-19 00:56:44
2022-11-19 00:57:58
2022-11-19 00:59:12
2022-11-19 01:00:26


In [53]:
len(per3)

4671

In [54]:
df3 = pd.DataFrame({'Time_Stamp': per3 })

# Creating data frame df4 23-11 to 26 -11 


In [48]:
import pandas as pd

per4 = pd.date_range(start ='11-23-2022', 
         end ='11-27-2022', freq ='103S')
  
for val in per4:
    print(val)

2022-11-23 00:00:00
2022-11-23 00:01:43
2022-11-23 00:03:26
2022-11-23 00:05:09
2022-11-23 00:06:52
2022-11-23 00:08:35
2022-11-23 00:10:18
2022-11-23 00:12:01
2022-11-23 00:13:44
2022-11-23 00:15:27
2022-11-23 00:17:10
2022-11-23 00:18:53
2022-11-23 00:20:36
2022-11-23 00:22:19
2022-11-23 00:24:02
2022-11-23 00:25:45
2022-11-23 00:27:28
2022-11-23 00:29:11
2022-11-23 00:30:54
2022-11-23 00:32:37
2022-11-23 00:34:20
2022-11-23 00:36:03
2022-11-23 00:37:46
2022-11-23 00:39:29
2022-11-23 00:41:12
2022-11-23 00:42:55
2022-11-23 00:44:38
2022-11-23 00:46:21
2022-11-23 00:48:04
2022-11-23 00:49:47
2022-11-23 00:51:30
2022-11-23 00:53:13
2022-11-23 00:54:56
2022-11-23 00:56:39
2022-11-23 00:58:22
2022-11-23 01:00:05
2022-11-23 01:01:48
2022-11-23 01:03:31
2022-11-23 01:05:14
2022-11-23 01:06:57
2022-11-23 01:08:40
2022-11-23 01:10:23
2022-11-23 01:12:06
2022-11-23 01:13:49
2022-11-23 01:15:32
2022-11-23 01:17:15
2022-11-23 01:18:58
2022-11-23 01:20:41
2022-11-23 01:22:24
2022-11-23 01:24:07


In [49]:
len(per4)

3356

In [50]:
df4 = pd.DataFrame({'Time_Stamp': per4 })

In [51]:
df4.tail

<bound method NDFrame.tail of               Time_Stamp
0    2022-11-23 00:00:00
1    2022-11-23 00:01:43
2    2022-11-23 00:03:26
3    2022-11-23 00:05:09
4    2022-11-23 00:06:52
...                  ...
3351 2022-11-26 23:52:33
3352 2022-11-26 23:54:16
3353 2022-11-26 23:55:59
3354 2022-11-26 23:57:42
3355 2022-11-26 23:59:25

[3356 rows x 1 columns]>

In [45]:
# df=df.append(df4, ignore_index = True)

C:\Users\pc\AppData\Local\Temp\ipykernel_150244\2512910618.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df=df.append(df4, ignore_index = True)


# Creating timestamp df5  27-11-2022 t0 30-11-2022

In [44]:
# create time stamp
# using pandas to_datetime() function

import pandas as pd

per5 = pd.date_range(start ='11-27-2022', 
         end ='12-1-2022', freq ='89S')
  
for val in per5:
    print(val)

2022-11-27 00:00:00
2022-11-27 00:01:29
2022-11-27 00:02:58
2022-11-27 00:04:27
2022-11-27 00:05:56
2022-11-27 00:07:25
2022-11-27 00:08:54
2022-11-27 00:10:23
2022-11-27 00:11:52
2022-11-27 00:13:21
2022-11-27 00:14:50
2022-11-27 00:16:19
2022-11-27 00:17:48
2022-11-27 00:19:17
2022-11-27 00:20:46
2022-11-27 00:22:15
2022-11-27 00:23:44
2022-11-27 00:25:13
2022-11-27 00:26:42
2022-11-27 00:28:11
2022-11-27 00:29:40
2022-11-27 00:31:09
2022-11-27 00:32:38
2022-11-27 00:34:07
2022-11-27 00:35:36
2022-11-27 00:37:05
2022-11-27 00:38:34
2022-11-27 00:40:03
2022-11-27 00:41:32
2022-11-27 00:43:01
2022-11-27 00:44:30
2022-11-27 00:45:59
2022-11-27 00:47:28
2022-11-27 00:48:57
2022-11-27 00:50:26
2022-11-27 00:51:55
2022-11-27 00:53:24
2022-11-27 00:54:53
2022-11-27 00:56:22
2022-11-27 00:57:51
2022-11-27 00:59:20
2022-11-27 01:00:49
2022-11-27 01:02:18
2022-11-27 01:03:47
2022-11-27 01:05:16
2022-11-27 01:06:45
2022-11-27 01:08:14
2022-11-27 01:09:43
2022-11-27 01:11:12
2022-11-27 01:12:41


In [45]:
len(per5)

3884

In [46]:
# create data frame
df5 = pd.DataFrame({'Time_Stamp': per5 })

In [49]:
# df=df.append(df5, ignore_index = True)

C:\Users\pc\AppData\Local\Temp\ipykernel_150244\3281976907.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df=df.append(df5, ignore_index = True)


# create df6 for future values 1-12- t0 3-12

In [38]:
# create time stamp
# using pandas to_datetime() function

import pandas as pd

per6 = pd.date_range(start ='12-1-2022', 
         end ='12-4-2022', freq ='103S')
  
for val in per6:
    print(val)

2022-12-01 00:00:00
2022-12-01 00:01:43
2022-12-01 00:03:26
2022-12-01 00:05:09
2022-12-01 00:06:52
2022-12-01 00:08:35
2022-12-01 00:10:18
2022-12-01 00:12:01
2022-12-01 00:13:44
2022-12-01 00:15:27
2022-12-01 00:17:10
2022-12-01 00:18:53
2022-12-01 00:20:36
2022-12-01 00:22:19
2022-12-01 00:24:02
2022-12-01 00:25:45
2022-12-01 00:27:28
2022-12-01 00:29:11
2022-12-01 00:30:54
2022-12-01 00:32:37
2022-12-01 00:34:20
2022-12-01 00:36:03
2022-12-01 00:37:46
2022-12-01 00:39:29
2022-12-01 00:41:12
2022-12-01 00:42:55
2022-12-01 00:44:38
2022-12-01 00:46:21
2022-12-01 00:48:04
2022-12-01 00:49:47
2022-12-01 00:51:30
2022-12-01 00:53:13
2022-12-01 00:54:56
2022-12-01 00:56:39
2022-12-01 00:58:22
2022-12-01 01:00:05
2022-12-01 01:01:48
2022-12-01 01:03:31
2022-12-01 01:05:14
2022-12-01 01:06:57
2022-12-01 01:08:40
2022-12-01 01:10:23
2022-12-01 01:12:06
2022-12-01 01:13:49
2022-12-01 01:15:32
2022-12-01 01:17:15
2022-12-01 01:18:58
2022-12-01 01:20:41
2022-12-01 01:22:24
2022-12-01 01:24:07


In [39]:
df6=pd.DataFrame({'Time_Stamp': per6 })

In [40]:
len(per6)

2517

# concatination

In [60]:
print(len(per1),len(per2),len(per3),len(per4),len(per5),len(per6))

2010 3563 4671 3356 3884 2517


In [61]:
# concatanate data frames
# empty data frame

df_merged=pd.DataFrame()
df_merged = pd.concat([df, df2,df3,df4,df5,df6], ignore_index = True)

In [62]:
df_merged.shape

(20001, 1)

<b>************************** 200001 timestamp records created ***********************************************</b>

# Creating 20001 enteries with event values

Search_click
Referral_click
added_to_basket
Product_sold
recommendation_click
saved_items

In [166]:
list=['search_click','referral_click','added_to_basket','product_sold','recommendation_click','saved_items']

In [167]:
df_List=pd.DataFrame(columns=['event_type'])
# for i in range(0,4333):
#     df_list=

In [168]:
# generate list and concatenate to populate the event_type column
df_List['event_type']=list*1321

In [169]:
min_List=pd.DataFrame(columns=['event_type'])
min_List['event_type']=list[0:5]*1921

In [170]:
mmin_List=pd.DataFrame(columns=['event_type'])
mmin_List['event_type']=list[0:2]*775

In [171]:
print(len(min_List),len(df_List),len(mmin_List))

9605 7926 1550


In [176]:
# rem_List an empty data frame
rem_List=pd.DataFrame(columns=['event_type'])
rem_List['event_type'] = ['product_sold' for x in range(920)]


In [177]:
print(len(min_List)+len(df_List)+len(mmin_List)+len(rem_List))

20001


In [181]:
# make a dataframe for all values of event_type
df_event_type=pd.DataFrame()
df_event_type=pd.concat([min_List,df_List,mmin_List,rem_List], ignore_index = True)
# find the unique values of event_type in number
df_event_type['event_type'].value_counts()
# df_event_type['event_type'].unique()


product_sold            4162
search_click            4017
referral_click          4017
added_to_basket         3242
recommendation_click    3242
saved_items             1321
Name: event_type, dtype: int64

In [159]:
rem_List.head()

event_type
0  product_sold
1  product_sold
2  product_sold
3  product_sold
4  product_sold

In [157]:
# empty data frame
import numpy as np
rem_List=pd.DataFrame(columns=['event_type'])
# len(rem_List)
vals = list[3]*850
ind = np.arange[len(vals)]

# df = pandas.DataFrame({'New column':vals})
rem_List=pd.DataFrame({'event_type':vals},index=ind)
# add values the event_type column with product_sold

# rem_List['event_type']=
# rem_List['event_type']=list[3]*850

TypeError: 'builtin_function_or_method' object is not subscriptable

# Take prduct Ids from shopgram table

In [186]:
import pymongo
import pandas as pd
# db_server=mongodb address of shopgram data 

db_server = "mongodb://mongoadmin:"
 
def save_to_db(df,db_server,db_name, collection, del_db = False):
	"""
	Saves a dataframe to a mongoDB collection
	Delets all the data in the collection if del_db = True
	"""
	client = pymongo.MongoClient(db_server)
	#for testing
	#importing database
	db  = client[db_name]
	if del_db != False:
		if collection in db.list_collection_names():
			# if db.collection_names().count('asiaOne_keywords') > 0:
			db[collection].delete_many({})
			
	data = df.to_dict(orient="records")
	db[collection].insert_many(data)
	client.close()


def read_from_db_by_date(db_server,db_name,collection,start_date,end_date):
	"""
	Reads from a mongoDB collection by date
    Parameters
    ----------
    db_server : str
        The mongoDB server to connect to
    collection : str    
        The collection to read from
    start_date : datetime    
        The start date to read from
    end_date : datetime  
        The end date to read from
    Returns
    -------
    df : dataframe
        The dataframe read from the collection
	"""
	client = pymongo.MongoClient(db_server)
	db  = client[db_name]
	#find the data filtered by the ga:date column with start date and end date in python 3
	#convert the dates to unix time with 00 hour 00 minute 00 second and 000000 microsecond
	start_date = int(start_date.timestamp()) *1000000000
	end_date = int(end_date.timestamp()) *1000000000
	#find the entries in teh mongo db that are between the start and end date
	data = db[collection].find({'ga:date':{'$gte':start_date,'$lte':end_date}})
	df = pd.DataFrame(list(data))
	#convert ga:date in df to datetime
	df['ga:date'] = pd.to_datetime(df['ga:date'])
	client.close()

	return df


def db_get_articles_by_url(db_server,db_name,collection,df):
    """
    This function takes a dataframe and returns a dataframe with the articles from the mongodb
    
    """
    client = pymongo.MongoClient(db_server)
    db  = client[db_name]
    #find the articles based on the route column of df 
    data = db[collection].find({'path':{'$in':df['path'].tolist()}})

    df = pd.DataFrame(list(data))

    #convert ga:date in df to datetime

    client.close()
    return df

#read all rows from the collection
def read_from_db(db_server,db_name, collection):
    """
    Reads from a mongoDB collection
    Parameters
    ----------
    db_server : str
        The mongoDB server to connect to
    collection : str
        The collection to read from
    Returns
    ------- 
    df : dataframe
        The dataframe read from the collection
    """
    client = pymongo.MongoClient(db_server)
    db  = client[db_name]
    data = db[collection].find()
    df = pd.DataFrame(list(data))
    client.close()
    return df
    
def db_check_not_empty(db_server,db_name,collection):
    """
    Checks if a collection is empty
    Parameters
    ----------
    db_server : str
        The mongoDB server to connect to
    collection : str    
        The collection to check
    Returns 
    ------- 
    """
    client = pymongo.MongoClient(db_server)
    db  = client[db_name]
    #data = db[collection].find()
    #if statement to check if the collection is empty, use count_documents to check if the collection is empty
    if db[collection].count_documents({}) == 0:
        return False
    else:
        return True

def db_get_newest_entry(db_server,db_name,collection):
    """
    Gets the newest entry in a collection
    Parameters
    ----------
    db_server : str
        The mongoDB server to connect to
    collection : str
        The collection to check
    Returns
    -------
    df : dataframe
        The newest entry in the collection
    """
    client = pymongo.MongoClient(db_server)
    db  = client[db_name]
    data = db[collection].find().sort('_id',-1).limit(1)
    df = pd.DataFrame(list(data))
    client.close()
    return df

In [1]:

db_server = "mongodb+srv://neurallab:neruallab@cluster0.mzwbp9n.mongodb.net/?retryWrites=true&w=majority"
df = read_from_db(db_server, "dummyDB","Shopgram_Data")
df.columns

NameError: name 'read_from_db' is not defined

In [2]:
import pymongo
import pandas as pd
#from mongoCloud_func import *
# from loc_recc import *
import requests

import json
import time
# import pymongo #Testing
from mongodb_functions import *
from nl_exception_handeling import *
import random
import string
# import pymongo
# from irefer_recc_eng import *
# Write internaldb_server address here
internal_db_server = "mongodb://mongoadmin:"
client = pymongo.MongoClient(internal_db_server)
internal_client=pymongo.MongoClient(internal_db_server)
# comment 23-36 for updating to  github
# "dummyDB","Shopgram_Data"
df_user = read_from_db( #this is the function to call
       client,'dummyDB',**{'db_name':'Shopgram_Data'})
    
# if len(df_user) == 0:
#         print('Empty user Data')
#         raise Exception('Empty user Data!!!')
# else:
#         print('Got user Data')


df_product = read_from_db(client,'Shopgram_Data',**{'db_name':'dummyDB'})

if len(df_product) == 0:
    print('Empty Product Data')
    raise Exception('Empty Product Data!!!')
else:
        print('Got Product Data')
        print(df_product.shape)


ModuleNotFoundError: No module named 'mongodb_functions'

In [191]:
df_product.tail()

_id    product_id  \
18439  635ec0da0f287670d4d7342f  6.587230e+12   
18440  635ec0da0f287670d4d73431  8.043740e+12   
18441  635ec0da0f287670d4d73439  6.762910e+12   
18442  635ec0da0f287670d4d73454  2.427670e+11   
18443  635ec0da0f287670d4d73470  5.744910e+12   

                                                   title   product_type  \
18439  Greeting Card - Alex Clark - 3 Year Old Boy - ...  Greeting Card   
18440                 EC Glitter Paint - 500ml -  Silver   Art Supplies   
18441  EC Liquitemp Classroom Poster Paint - 2 Litre ...   Art Supplies   
18442           MAGFORMERS - Accessory - Gears - Pack 20   Construction   
18443   Coral Kingdom - Australian Geographic - Hardback           BOOK   

                vendor                                               tags  
18439         Waterlyn         3+,greeting cards,spo-default,spo-disabled  
18440              GNS  2+,3+,4 +,5+,6+,7+,8+,Art & Craft,Art Supplies...  
18441              GNS  2+,3+,4+,5+,6+,7+,8+,Art & Craft,Art Supplies,...  
18442     lifes little  3+,4 +,5+,6+,Boys,Construction,Construction & ...  
18443  Brumby Sunstate        3+,4+,5+,6+,books,Boys,Girls,Picture,Unisex

In [211]:
min_product_df=df_product.iloc[15000:16557]

In [212]:
min_product_df.shape

(1557, 6)

In [213]:
df_product.shape

(18444, 6)

In [214]:
df_product.shape[0]+min_product_df.shape[0]

20001

In [215]:
newdf=pd.concat([df_product,min_product_df])

In [216]:
newdf.shape

(20001, 6)

In [217]:
newdf.tail

<bound method NDFrame.tail of                             _id    product_id  \
0      635ebda9f4aa708967de9d75  3.130210e+11   
1      635ebda9f4aa708967de9d7b  3.130220e+11   
2      635ebda9f4aa708967de9d7d  5.289110e+11   
3      635ebda9f4aa708967de9d82  5.289230e+11   
4      635ebda9f4aa708967de9d9a  7.371600e+11   
...                         ...           ...   
16552  635ec0da0f287670d4d72def  4.581500e+12   
16553  635ec0da0f287670d4d72e00  4.570550e+12   
16554  635ec0da0f287670d4d72e01  1.947090e+11   
16555  635ec0da0f287670d4d72e07  4.570540e+12   
16556  635ec0da0f287670d4d72e0c  2.185506e+09   

                                                   title   product_type  \
0                                     JinMing M4A1 Gen 8          Rifle   
1                                          Tactical Vest          Vests   
2           Single Point Tactical Sling with Bungee Cord    Accessories   
3                                       Magazine Coupler    Accessories   
4     

In [218]:
# reset index of the dataframe
newdf.reset_index(drop=True, inplace=True)


In [219]:
newdf.tail()

_id    product_id  \
19996  635ec0da0f287670d4d72def  4.581500e+12   
19997  635ec0da0f287670d4d72e00  4.570550e+12   
19998  635ec0da0f287670d4d72e01  1.947090e+11   
19999  635ec0da0f287670d4d72e07  4.570540e+12   
20000  635ec0da0f287670d4d72e0c  2.185506e+09   

                                                   title   product_type  \
19996  D&D Icewind Dale Rime of the Frostmaiden Ice C...      Miniature   
19997                     Jump Around Frogs - Wetthupfen    Board Games   
19998           Summoner's Art - AP06-EN011 - Super Rare    yugio cards   
19999  My Very First Games – Building Site - Meine ...    Board Games   
20000              Abyss Dweller - THSF-EN047 Super Rare  Trading Cards   

                     vendor                                              tags  
19996  Wizards of the Coast                               Dungeon and Dragons  
19997                  HABA  aug,Board Games,date-30-08-2021,OzSale,pre-order  
19998                Konami                 Catch,Kogan,OzSale,Yugioh Singles  
19999                  HABA  Board Games,date-30-11-2021,nov,OzSale,pre-order  
20000                Konami                                   Catch,Kogan,XYZ

# Generating user ids -->

init='63'
# generate random combination of digits and alphabets

In [316]:
init='634'
# generate random combination of digits and alphabets
# of length 23
Greatres=[]
for i in range(6001):
    res = init + ''.join(random.choices(string.ascii_uppercase +
                             string.digits, k = 23))
    Greatres.append(res)
# res = init + ''.join(random.choices(string.ascii_uppercase +string.ascii_lowercase + string.digits, k = 22))

In [317]:
print(len(Greatres))

6001


In [318]:
df_dummyuser=pd.DataFrame(Greatres,columns=['user_id'])

In [319]:
df_dummyuser.head()

user_id
0  634216K2JNI0KBF5ABMG53DXY9
1  634D5PDN19MCS1N2ZS7M8VC06X
2  634E8DRXJCH307R79LMW02G0L8
3  634IHVAKDEON7IL8NZH7XE22MC
4  6346P66BXHOW8XH3FVVV3IQ750

In [320]:
# duplicate values of dummy user
init='633'
# generate random combination of digits and alphabets
# of length 23
Greatres=[]
for i in range(1000):
    res = init + ''.join(random.choices(string.ascii_uppercase +
                             string.digits, k = 23))
    Greatres.append(res)

In [321]:
df_dummyuser1=pd.DataFrame(Greatres,columns=['user_id'])

In [322]:
# concat the two dummy user
df_dummyuser=pd.concat([df_dummyuser,df_dummyuser1])

In [323]:
df_dummyuser.shape

(7001, 1)

In [324]:
# shuffle the dummy user
df_dummyuser=df_dummyuser.sample(frac=1).reset_index(drop=True)

In [325]:
# trim slices of dummy users
trim1_df_dummyuser=df_dummyuser.iloc[0:1000]

In [326]:
# shuffle the dummy user
df_dummyuser=df_dummyuser.sample(frac=1).reset_index(drop=True)

In [327]:

# concat with the trimmed dummy user
df_dummyuser=pd.concat([df_dummyuser,trim1_df_dummyuser])

In [328]:
df_dummyuser.shape

(8001, 1)

In [329]:
for i in range(12):
    df_dummyuser['user_id'][i]='633'+df_dummyuser['user_id'][i]
    df_dummyuser=df_dummyuser.sample(frac=1).reset_index(drop=True)
    trim1_df_dummyuser=df_dummyuser.iloc[0:1000]
    df_dummyuser=df_dummyuser.sample(frac=1).reset_index(drop=True)
    df_dummyuser=pd.concat([df_dummyuser,trim1_df_dummyuser])

In [330]:
df_dummyuser.shape

(20001, 1)

In [331]:
df_dummyuser.tail()

user_id
995  634MNS0H7Y9DVNT2ET8EP6QBEB
996  634PXFXNM0CLMEAPYDB74JNUXI
997  634LWTMBQGDXSN9AQ4R69BRH9B
998  634AM0RHQQINHKJ3QE8Y9MB35C
999  634O3ER5ST9LSXDS95IG1SZ9JH

In [287]:
df_dummyuser.value_counts()

user_id                   
633YDEFD5B0F3H17DFPRG27PIW    20
633G0QEY5XWU5YRAUXFKLFD8IO    19
63496Z4MFQ3Z6KGDPZVFIRHKSD    17
634J9JO6HH24DHXDNQGD3BS8P4    17
633WGJ43WJ51GKDVBZOHP7EGCD    14
                              ..
6347I7DNI5V6L6PJ835Q9VYJIL     1
6347HGG0AGCHGD4RLLG4FMYZBN     1
6347G9WTZXSGW6J99SK08LJ4FS     1
6347G9FDQSYR5M2COYI0A04M1G     1
634EL9ZFZBVAWZ10KXO5GBW79R     1
Length: 7020, dtype: int64

# Required dataframes created

In [295]:
df_event_type.shape

(20001, 1)

In [296]:
# product dataframe
newdf.shape
# newdf.head()

(20001, 6)

In [297]:
# Timestamp
df_merged.shape
# df_merged.head()

(20001, 1)

In [332]:
# user dataframe
df_dummyuser.shape


(20001, 1)

# Generate 4 columns from 4 different dataframes and create one dataframe

In [333]:
# generate four different columns from df_event_type, df_merged, df_dummyuser, newdf
# df_event_type
# df_merged
# df_dummyuser
# newdf
finaldf=pd.DataFrame(columns=['user_id','product_id','event_type','event_timestamp'])

In [334]:
# finaldf first column= df_dummyuser
finaldf['user_id']=df_dummyuser['user_id']

In [335]:
finaldf.head()

user_id product_id event_type event_timestamp
0  634HEZTVCB8KVEA6QG4D7AKRHN        NaN        NaN             NaN
1  6347SX8QVIYGS6KQM920LYA5PH        NaN        NaN             NaN
2  633ZWLCP4DFQC0V8ZGZ5KNV0GA        NaN        NaN             NaN
3  634PHK1EMAD21DR5CLINUV39XQ        NaN        NaN             NaN
4  6349FL7J8ZW7T4VT92M4GPI4HF        NaN        NaN             NaN

In [336]:
# finaldf second column= newdf
finaldf['product_id']=newdf['product_id']
# finaldf['user_id']=df_dummyuser['user_id']

In [337]:
finaldf.head()

user_id    product_id event_type event_timestamp
0  634HEZTVCB8KVEA6QG4D7AKRHN  3.130210e+11        NaN             NaN
1  6347SX8QVIYGS6KQM920LYA5PH  3.130220e+11        NaN             NaN
2  633ZWLCP4DFQC0V8ZGZ5KNV0GA  5.289110e+11        NaN             NaN
3  634PHK1EMAD21DR5CLINUV39XQ  5.289230e+11        NaN             NaN
4  6349FL7J8ZW7T4VT92M4GPI4HF  7.371600e+11        NaN             NaN

In [338]:
# finaldf third column= df_event_type
finaldf['event_type']=df_event_type['event_type']
# finaldf['product_id']=newdf['product_id']
# finaldf['user_id']=df_dummyuser['user_id']

In [339]:
finaldf.head()

user_id    product_id            event_type  \
0  634HEZTVCB8KVEA6QG4D7AKRHN  3.130210e+11          search_click   
1  6347SX8QVIYGS6KQM920LYA5PH  3.130220e+11        referral_click   
2  633ZWLCP4DFQC0V8ZGZ5KNV0GA  5.289110e+11       added_to_basket   
3  634PHK1EMAD21DR5CLINUV39XQ  5.289230e+11          product_sold   
4  6349FL7J8ZW7T4VT92M4GPI4HF  7.371600e+11  recommendation_click   

  event_timestamp  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN

In [343]:
# finaldf fourth column= df_merged

finaldf['event_timestamp']=df_merged['Time_Stamp']
# finaldf['event_type']=df_event_type['event_type']
# finaldf['product_id']=newdf['product_id']
# finaldf['user_id']=df_dummyuser['user_id']

In [344]:
finaldf.shape

(20001, 4)

In [347]:
finaldf.tail()

user_id    product_id            event_type  \
995  634MNS0H7Y9DVNT2ET8EP6QBEB  4.856160e+12          search_click   
996  634PXFXNM0CLMEAPYDB74JNUXI  4.864640e+12        referral_click   
997  634LWTMBQGDXSN9AQ4R69BRH9B  4.879810e+12       added_to_basket   
998  634AM0RHQQINHKJ3QE8Y9MB35C  4.879820e+12          product_sold   
999  634O3ER5ST9LSXDS95IG1SZ9JH  4.914920e+12  recommendation_click   

        event_timestamp  
995 2022-11-12 11:25:25  
996 2022-11-12 11:29:00  
997 2022-11-12 11:32:35  
998 2022-11-12 11:36:10  
999 2022-11-12 11:39:45

In [356]:
# reset index of the dataframe bring right order
# finalDf=finaldf
finalDf=finalDf.reset_index(drop=True, inplace=False)


In [357]:
# save the dataframe
finalDf.to_csv('finalDf.csv')

In [352]:
# save the dataframe 
finalDf.to_csv('finalDf.csv')

In [355]:
finalDf.to_csv('finalDf.csv')

In [359]:
df_event_type.value_counts()

event_type          
product_sold            4162
referral_click          4017
search_click            4017
added_to_basket         3242
recommendation_click    3242
saved_items             1321
dtype: int64